gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExplicitSourceCompatibilityTask.java

```diff
@@ -0,0 +1,110 @@
+                + " on $JAVA_HOME which is fragile.");
+        this.shouldFix = objectFactory.property(Boolean.class);
+        this.shouldFix.set(false);
+
+        onlyIf(new Spec<Task>() {
+            @Override
+            public boolean isSatisfiedBy(Task element) {
+                // sometimes people apply the 'java' plugin to projects that doesn't actually have any java code in it
+                // (e.g. the root project), so if they're not publishing anything, then we don't bother enforcing the
+                // sourceCompat thing
+                PublishingExtension publishing = getProject().getExtensions().findByType(PublishingExtension.class);
+                return publishing != null;
+            }
+        });
+    }
+
+    @Option(option = "fix", description = "Whether to apply the suggested fix to build.gradle")
+    public final void setShouldFix(boolean value) {
+        shouldFix.set(value);
+    }
+
+    @TaskAction
+    public final void taskAction() throws IOException {
+        // We're doing this naughty casting because we need access to the `getRawSourceCompatibility` method.
+        org.gradle.api.plugins.internal.DefaultJavaPluginConvention convention =
+                (org.gradle.api.plugins.internal.DefaultJavaPluginConvention)
+                        getProject().getConvention().getPlugin(JavaPluginConvention.class);
+
+        if (convention.getRawSourceCompatibility() != null) {
+            // In theory, users could configure the fancy new 'java toolchain' as an alternative to explicit
+            // sourceCompatibility, but there's no method to access this yet (as of Gradle 6.8).
+            return;
+        }
+
+        if (shouldFix.get()) {
+            Files.write(
+                    getProject().getBuildFile().toPath(),
+                    Collections.singletonList(String.format("%nsourceCompatibility = %s", JavaVersion.current())),
+                    StandardCharsets.UTF_8,
+                    StandardOpenOption.APPEND,
+                    StandardOpenOption.CREATE);
+            return;
+        }
+
+        throw new GradleException(String.format(
+                "%s must set sourceCompatibility explicitly in '%s', "
+                        + "otherwise compilation will not be reproducible but instead depends on the Java version "
+                        + "that Gradle is currently running with (%s). To auto-fix, run%n"
+                        + "%n"
+                        + "     ./gradlew %s --fix%n"
+                        + "%n"
+                        + "This will automatically add a suggested line "
+                        + "(you may need to adjust the number, e.g. to '1.8' for maximum compatibility).",
+                getProject(),
+                getProject().getRootProject().relativePath(getProject().getBuildFile()),
+                JavaVersion.current(),
+                getPath()));
```

Comment:
I broke tracing-java and tried it out, error message looks like this:
```
1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':tracing-jaxrs:checkExplicitSourceCompatibility'.
> project ':tracing-jaxrs' must set sourceCompatibility explicitly in 'tracing-jaxrs/build.gradle', otherwise compilation will not be reproducible but instead depends on the Java version that Gradle is currently running with (11). To auto-fix, run

       ./gradlew :tracing-jaxrs:checkExplicitSourceCompatibility --fix

  This will automatically add a suggested line (you may need to adjust the number, e.g. to '1.8' for maximum compatibility).

```